Add proto to sockets

carson-coder opened the pull request:
Add #8750

carson-coder commented:
espressif port is tested but I couldn't test the raspberry pi because I don't have a device to test it

carson-coder commented:
looks like raspberrypi_pico and some other builds failed so I will try to fix that

jepler reviewed (changes requested):
This commit adds some files that should probably not have been added (os, os.path, etc) with many thousands of lines of content. Please **rebase** this branch so that these files are not added to our history. If that's beyond your git skills, I can help you figure out how.

The rest looks plausible, based on my small understanding of networking.

carson-coder commented on shared-bindings/socketpool/SocketPool.c:
```diff
@@ -86,21 +86,25 @@ MP_DEFINE_EXCEPTION(gaierror, OSError)
 //|
 //|         :param ~int family: AF_INET or AF_INET6
 //|         :param ~int type: SOCK_STREAM, SOCK_DGRAM or SOCK_RAW
+//|         :param ~int proto: Int
```
Ok. I can work on this tomorrow.

carson-coder commented:
Could you help me with git rebase @jepler

dhalbert commented:
Do you have `adafruit/circuitpython as a remote, e.g.:
```
$ git remote -v
adafruit	https://github.com/adafruit/circuitpython.git (fetch)
adafruit	https://github.com/adafruit/circuitpython.git (push)
origin	https://github.com/dhalbert/circuitpython.git (fetch)
origin	https://github.com/dhalbert/circuitpython.git (push)
```
It might be called `upstream` instead. If not add it as a remote.
So, to rebase:
```
$ git fetch adafruit
$ git rebase adafruit/main
```

carson-coder commented:
> ```
> $ git fetch adafruit
> $ git rebase adafruit/main
> ```

Could I do adafruit/8.2.x because I am adding this on the 8.2.x branch

carson-coder commented:
I think I did it

dhalbert commented:
> ould I do adafruit/8.2.x because I am adding this on the 8.2.x branch

fer sure, and I guess you did

carson-coder commented:
Yep

jepler commented:
Looks like pre-commit wants a few more changes. Other than that, this looks good to me (but i didn't test)

dhalbert commented:
What kind of network testing do you want to do with this? Can you write a ping?

300bps commented on shared-bindings/socketpool/SocketPool.c:
```diff
@@ -86,21 +86,25 @@ MP_DEFINE_EXCEPTION(gaierror, OSError)
 //|
 //|         :param ~int family: AF_INET or AF_INET6
 //|         :param ~int type: SOCK_STREAM, SOCK_DGRAM or SOCK_RAW
+//|         :param ~int proto: IPPROTO_TCP or IPPROTO_IP. Only works with SOCK_RAW 
```
I manually passed the value of proto = 1 for ICMP and it works to allow me to create a manual ICMP packet with a raw socket (I've tested it).  Unfortunately, it doesn't look like IPPROTO_ICMP = 1 is defined anywhere, in the existing code base.

300bps commented:
I've been testing the Raspberry Pi Pico version of this build (some comments were in the associated issue).  The results are:
1) These code changes in this build SUCCESSFULLY allow me to create a raw ICMP (ping) socket by passing proto=1. Previously, this was not supported.
```
    import socketpool
    import wifi

    IP_HEADER_PROTOCOL_ICMP = 1
    sp = socketpool.SocketPool(wifi.radio)
    sp.socket(family=socketpool.AF_INET, type=socketpool.SOCK_RAW, proto=IP_HEADER_PROTOCOL_ICMP)
```
2) I've manually created ICMP ping (echo) raw packets and SUCCESSFULLY sent them and received ping responses from a number of sites: (1.1.1.1, 8.8.8.8, 1.0.0.1, and 8.8.4.4, along with my local router).  I've tested it by continuously pinging once per second for several hours at a time.  I've also run faster rapid-fire pings for shorter intervals successfully.
3) I checked the circuitpython codebase and didn't find an IPPROTO_ICMP = 1 or SOCKETPOOL_IPPROTO_ICMP = 1 that could be passed as a symbolic constant for the 'proto' parameter in the 'socket' call.  This isn't necessary for the changes to be usable and functional.  If this were desired to be added, the locations could be found by searching 'IPPROTO_IP' or 'IPPROTO_TCP'.
4) The only unexpected behavior that I've observed is: if I attempt to open 2 or more ICMP raw sockets, the last one opened functions correctly and the earlier ones repeatedly time out. I don't see a direct cause of that in the changes in this pull request, and think it seems most likely to be a consequence or limitation of the lower-level 'lwip' networking code.
5) All-in-all the changes in this PR do enable the ability to use raw sockets to build ICMP and other raw messages and it brings CP more in-line with regular python socket functionality.  I recommend adopting it into the main for new CP9 builds.

Kudos @carson-coder for enabling this functionality.

300bps commented:
> What kind of network testing do you want to do with this? Can you write a ping?

@dhalbert I did successfully do this with changes in this PR for the RPi Pico and it works well.  See my previous comment above for more detailed results.  Thanks!

anecdata commented:
I think `socketpool.SocketPool.IPPROTO_ICMP` could be added:
```
>>> help(socketpool.SocketPool)
object <class 'SocketPool'> is of type type
  socket -- <function>
  getaddrinfo -- <function>
  gaierror -- <class 'gaierror'>
  AF_INET -- 2
  AF_INET6 -- 10
  SOCK_STREAM -- 1
  SOCK_DGRAM -- 2
  SOCK_RAW -- 3
  TCP_NODELAY -- 1
  IPPROTO_TCP -- 6
  IPPROTO_IP -- 0
  IP_MULTICAST_TTL -- 5
  EAI_NONAME -- -2
```
https://github.com/adafruit/circuitpython/blob/6627bd7826fc7bbcddb33dae069d672d2d663af8/shared-bindings/socketpool/SocketPool.h#L47

carson-coder commented:
For constantant for the proto parm I just use the Two that already exist

carson-coder commented:
Should I add every IPPROTO_* constant. And what would I do for the `socketpool.socket()` docstring because listing every constant isn't viable because there are a lot

dhalbert commented:
You could add only the ones that are actually implemented by Espressif or RP2040 wifi. And yes, add each one, with a docstring underneath. as class constants.

carson-coder commented:
I just added each one and in the docstring for socket I said that proto can be `IPPROTO_TCP, IPPROTO_IP or any other IPPROTO_ constant.` and I added each constant in the class doc string. you can look at the next commit.

dhalbert commented on shared-bindings/socketpool/SocketPool.c:
```diff
@@ -83,6 +83,30 @@ MP_DEFINE_EXCEPTION(gaierror, OSError)
 //|
 //|     IPPROTO_TCP: int
 //|     IPPROTO_IP: int
+//|     IPPROTO_AH: int
+//|     IPPROTO_DSTOPTS: int
+//|     IPPROTO_EGP: int
+//|     IPPROTO_ESP: int
+//|     IPPROTO_FRAGMENT: int
+//|     IPPROTO_GRE: int
+//|     IPPROTO_HOPOPTS: int
+//|     IPPROTO_ICMP: int
+//|     IPPROTO_ICMPV6: int
+//|     IPPROTO_IDP: int
+//|     IPPROTO_IGMP: int
+//|     IPPROTO_IPIP: int
+//|     IPPROTO_IPV6: int
+//|     IPPROTO_MPTCP: int
+//|     IPPROTO_NONE: int
+//|     IPPROTO_PIM: int
+//|     IPPROTO_PUP: int
+//|     IPPROTO_RAW: int
+//|     IPPROTO_ROUTING: int
+//|     IPPROTO_RSVP: int
+//|     IPPROTO_SCTP: int
+//|     IPPROTO_TP: int
+//|     IPPROTO_UDP: int
+//|     IPPROTO_UDPLITE: int
```
Adding them here does not make them available in Python You need to add them around line 221 in this file. But there are too many to add, and most will not be used. Just add a few of the common ones, including ICMP.

For instance, MicroPython defines these common ones:
https://github.com/micropython/micropython/blob/9feb0689eeaca5ce88aedcc680f997a3b4d0221c/extmod/modsocket.c#L640

300bps replied:
IPPROTO_ICMP=1 and IPPROTO_RAW=255 are the 2 that'd I'd recommend explicitly adding to those that are already implemented in CP.

carson-coder commented:
CPython's socket doesn't define IPPROTO_IPV4 but the micropython file that was linked did so I am not going to define it because it is not in python

jepler reviewed (approved):
I think it's good to merge if the CI comes up green. Thanks for taking care of all the details here!

carson-coder commented:
Np

300bps commented:
Nice work!  

@dhalbert GitHub newbie question here, but it looks like this was merged into a branch (8.2.x). How does it get into main so that it will be available in the CP9.0 codebase?

dhalbert commented:
That was done here: #8769 by @jepler.

300bps commented:
> That was done here: #8769 by @jepler.

Awesome!  Thanks for the 411.
